src/cargo/ops/cargo_compile.rs

```diff
@@ -377,9 +386,19 @@ fn generate_targets<'a>(pkg: &'a Package,
                 try!(find(tests, "test", TargetKind::Test, test));
                 try!(find(benches, "bench", TargetKind::Bench, &profiles.bench));
             }
-            Ok(targets)
+            targets
         }
-    }
+    };
+
+    targets.retain(|&(t, _)| {
+        t.is_lib() ||
+        t.features().is_none() ||
+        t.features().unwrap().iter().filter(|&f| {
+            features.map_or(true, |x| !x.contains(f))
+        }).count() == 0
```

Comment:
It probably doesn't make sense to have lib targets with features, so can that be prevented during decoding of a manifest? That means this could look like:

```rust
match t.features() {
    Some(f) => f.iter().any(|f| features.contains(f))
    None => true,
}
```

(or something like that).

In general I'd recommend avoiding patterns like:

* `foo.is_some() || foo.unwrap().bar()` -- try to do the `match` only once to avoid `unwrap()`
* `iter().filter(..).count() == 0` is the same as leveraging `any()` or `all()` I believe
* `map_or` -- currently in Cargo I tend to prefer `.map().unwrap_or()` as the ordering of the arguments is a bit more natural